### PR TITLE
Upstream remaining shape functions from Torch-MLIR.

### PR DESCRIPTION
**Pull request by silvasean**
Follow-on to https://github.com/pytorch/pytorch/pull/76592 adding the
rest.

### CODE REVIEW

**Comment by silvasean** [comment]
cc @eellison @makslevental

**Comment by Gamrix** [comment]
@silvasean I'm looking at it.

**Comment by Gamrix** [comment]
@silvasean @eellison  Due to how we parse the shape functions for use, for functions that return shapes of multiple tensors, there must be only one `return` statement and it must be a tuple constructor.

File: torch/jit/_shape_functions.py [hunk 1]
```diff
@@ -860,6 +860,61 @@ def argmax(self: List[int], dim: Optional[int] = None, keepdim: bool = False) ->
...
+    for i in range(num_unreduced_dimensions):
+        reduction_shape.append(input[i])
+    for i in range(num_unreduced_dimensions, len(input)):
+        reduction_shape.append(1)
+    return input, reduction_shape, reduction_shape
+
+def native_batch_norm(input: List[int], weight: Optional[List[int]], bias: Optional[List[int]], running_mean: Optional[List[int]], running_var: Optional[List[int]], training: bool) -> Tuple[List[int], List[int], List[int]]:
+    if training:
+        return input, [input[1]], [input[1]]
+    return input, [0], [0]
```
Comment (Gamrix) [inline torch/jit/_shape_functions.py:908]: See main comment

```suggestion
    if training:
        _size = [input[1]]
    else:
        _size = [0]
    return input, _size, _size
```

File: torch/jit/_shape_functions.py [hunk 2]
```diff
@@ -860,6 +860,61 @@ def argmax(self: List[int], dim: Optional[int] = None, keepdim: bool = False) ->
...
+    assert target_dim <= 1
+    no_batch_dim = self_dim == 1 and target_dim == 0
+    assert no_batch_dim or (self[0] == target[0])
+    n_classes = self[-1]
+    scalar_shape: List[int] = []
+    assert weight is None or (len(weight) == 1 and weight[0] == n_classes)
+    if reduction == 0 and self_dim == 2:
+        return [self[0]], scalar_shape
+    else:
+        return scalar_shape, scalar_shape
```
Comment (Gamrix) [inline torch/jit/_shape_functions.py:893]:
```suggestion
    if reduction == 0 and self_dim == 2:
        reduction_shape = [self[0]]
    else:
        reduction_shape = scalar_shape
    return reduction_shape, scalar_shape
```

**Comment by eellison** [comment]
@Gamrix we can fix that by adding a tuple unpack node instead of this invariant https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/symbolic_shape_registry.cpp#L229

**Comment by silvasean** [comment]
It looks like the shape function for `aten::index.Tensor` is choking up the pass:

```
def index_Tensor(self: List[int], indices: List[Optional[List[int]]]) -> List[int]:
    assert len(indices) <= len(self), "More indices than dimensions to index"
    broadcasted_shape: List[int] = []
    for index_tensor_shape in indices:
        if index_tensor_shape is not None:
            broadcasted_shape = broadcast(broadcasted_shape, index_tensor_shape)
    return broadcasted_shape
```

I didn't see from the code an obvious way to handle the `List[Optional[List[int]]]` -- it looks like the code for `List[List[int]]` in `aten::cat` required a fair amount of special handling, so I'll leave it commented out for now.

**Comment by eellison** [comment]
Sounds good ! I think @gamrix was looking into generalizing that

**Comment by silvasean** [comment]
Do you folks see any legitimate CI failures left? The two left (see below) don't seem related).

Any insight on this CI failure? (`pull / linux-xenial-py3.7-gcc5.4 / test (backwards_compat, 1, 1, linux.2xlarge) (pull_request) `)

```
The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not. 

Broken ops: [
	aten::glu_jvp(Tensor glu, Tensor x, Tensor dx, int dim) -> (Tensor)
	aten::_resize_output_(Tensor(a!) self, int[] size, Device device) -> (Tensor(a!))
]
```


Also the `pull / pytorch-xla-linux-bionic-py3.7-clang8 / test (xla, 1, 1, linux.2xlarge) (pull_request) ` failure seems unrelated?

```
2022-05-12T16:12:30.2188210Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/tensor_impl.h:33:40: error: only virtual member functions can be marked 'override'
2022-05-12T16:12:30.2188582Z   at::IntArrayRef sizes_custom() const override;
2022-05-12T16:12:30.2188818Z                                        ^~~~~~~~
```

**Review by eellison — APPROVED** [review]
LGTM!! Thanks!!

**Pull request closed without merging**